pyk/src/pyk/kcfg/kcfg.py

```diff
@@ -584,10 +586,20 @@ def log(message: str, *, warning: bool = False) -> None:
                 log(f'abstraction node: {node.id}')
 
             case Step(cterm, depth, next_node_logs, rule_labels, _):
-                next_node = self.create_node(cterm)
-                logs[next_node.id] = next_node_logs
-                self.create_edge(node.id, next_node.id, depth, rules=rule_labels)
-                log(f'basic block at depth {depth}: {node.id} --> {next_node.id}')
+                if optimize_kcfg and (len(edges := self.edges(target_id=node.id)) == 1):
+                    in_edge = edges[0]
+                    self.remove_edge(in_edge.source.id, node.id)
+                    self.let_node(node_id=node.id, cterm=cterm)
+                    self.create_edge(
+                        in_edge.source.id, node.id, in_edge.depth + depth, list(in_edge.rules) + rule_labels
+                    )
+                    logs[node.id] = logs[node.id] + next_node_logs if node.id in logs else next_node_logs
+                    log(f'basic block at depth {depth}: update: {node.id}')
+                else:
+                    next_node = self.create_node(cterm)
+                    logs[next_node.id] = next_node_logs
+                    self.create_edge(node.id, next_node.id, depth, rules=rule_labels)
+                    log(f'basic block at depth {depth}: {node.id} --> {next_node.id}')
```

Comment:
I don't think we should re-use the original node-id actually, because it may be referred to by another edge (for example, a back-edge/cover that points at it, or part of a `MergedEdge`). So we should still generate a fresh new node that gets put into the KCFG (which would be common logic between the minimize branch and non-minimize branch).

We should probably also have a stronger check on the node, that it _only_ has the one incoming edge. Something like:

```suggestion
                next_node = self.create_node(cterm)
                predecessors = self.predecessors(node.id)
                if optimize_kcfg and len(predecessors) == 1 and isinstance(predecessors[0], KCFG.Edge):
                    in_edge = predecessors[0]
                    self.remove_edge(in_edge.source.id, node.id)
                    self.create_edge(
                        in_edge.source.id, next_node.id, in_edge.depth + depth, list(in_edge.rules) + rule_labels
                    )
                    logs[node.id] = logs[node.id] + next_node_logs if node.id in logs else next_node_logs
                    log(f'basic block at depth {depth}: extend {in_edge.source.id} --> {node.id} --> {next_node.id}')
                else:
                    logs[next_node.id] = next_node_logs
                    self.create_edge(node.id, next_node.id, depth, rules=rule_labels)
                    log(f'basic block at depth {depth}: {node.id} --> {next_node.id}')
```

Reply:
Yes, that is a good point. 

I think that we can only extend the KCFG from a leaf node, is that correct? Can we extend then even from a node that's been covered or a node that's part of a `MergedEdge`, or a node that has a back edge into itself (How could this happen to begin with, please remind me?).

It also seems to me that if we guard against the cases that you mentioned, I think that we can keep the node identifier safely, can we not?

Am I missing something?

Reply:
There must be some invariant on the predecessors, right? They can't all exist at the same time, surely...

Reply:
The only thing we guarantee about edges is that each one has at most one successor, and that no zero-length loops are introduced (for example, a split into a cover). So I think you could have multiple predecessors.

But more than that, we should just keep good hygiene with the identifiers and keep them unique, because they can be referenced by things higher up the stack (like the `terminal` nodes list, or the `bounded` nodes list, for example).

Reply:
Yes, I agree, that's fair.